### ISSUE TITLE
Bug: [no-unnecessary-condition] doesn't work with nullish coalescing operators

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#8079]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3gN1gRgC55MYBLFAc3gB94UBXCCAbgFgAoD3GPeAfj51GLDh1CRYCZOizcATERLkqteilAAzciGBtO7brwHw1m7bo5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gdQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const var1: string | null;

var1 ?? null;

declare const var2: string | undefined;

var1 ?? undefined;
```

### ESLint Config

```javascript
.
```

### Expected Result

I expect the rule to flag this code, but it doesn't.

### Actual Result

.

**Comment by Josh-Cena** [comment]
This is not an "unnecessary condition", as far as the rule is concerned. The rule reports cases where the RHS would never be evaluated, not when the RHS doesn't change the value of the expression. This does not error either:

```ts
declare const f: boolean;
f || false;
```

Nor this:

```ts
declare const f: number;
f || 0;
```

We can probably solve it in this case, but it's just a hard problem to solve in general: what about `f + 0`? That's also a useless identity!

**Comment by Zamiell** [comment]
My apologies, should it be a new rule then instead of a bug-fix for no-unnecessary-condition?
Is there some other nullish coalescing operator rule?

**Comment by Josh-Cena** [comment]
I don't think this should be specific to nullish coalescing? Surely `||`/`&&` could be supported too. As far as I know there's no rule to catch this either.

**Comment by Zamiell** [comment]
Ok. Brainstorming names: `no-useless-short-circuit` ?

Test cases:

```ts
declare const a: boolean;
f || false;

declare const a: number;
f || 0;

declare const a: string;
f || "";

declare const a: T | undefined;
f ?? undefined;

declare const a: T | null;
f ?? null;
```

**Comment by Zamiell** [comment]
> We can probably build these cases into the rule

I agree. After thinking on this more, I think these cases should be built into the existing rule. Take the following two examples:

```ts
declare const foo: string;
const bar = var1 ?? null;

declare const foo: string | null;
const bar = var1 ?? null;
```

Consider that:

1. These code snippets are almost the exact same thing
2. Fundamentally, short-circuit operators are just comparisons (which is why the existing rule has logic for them).
3. By any sane definition of the phrase "unnecessary comparison", both of these examples contain unnecessary comparisons.

Thus, I think it would be weird to suggest that the first bug should be inside of the scope of `no-unnecessary-condition`, but the second bug should be outside the scope and handled by some other rule.

**Comment by Zamiell** [comment]
Actually, I decided on the name `no-unnecessary-assignment` so that it would closer match `no-unnecessary-condition`.

I made the rule here: https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/docs/rules/no-unnecessary-assignment.md

@Josh-Cena Are there any more test cases you can think of that I should add? Also do you think I should submit this to `typescript-eslint`?

**Comment by bradzacher** [comment]
I'm okay with having this as a check within `no-unnecessary-condition`. It seems like a logical extension and the rule does already check those operators. Seems like a win to me and I don't know why you'd want this sort of dead code in your codebase.

**Comment by Zamiell** [comment]
Brad, there's logic in the rule to prevent this:

```ts
declare let foo: number;
foo += 0;
```

I don't think this check can be shoehorned into `no-unnecessary-condition`. Thus, having a new rule called `no-unnecessary-assignment` for a better delineation of concerns seems more logical.

**Comment by bradzacher** [comment]
But it's not just about assignments, right? It's just about the general pattern of a RHS value that equals the negative case (for `||` and `??`) or the positive case (for `&&`). So we want to apply this logic to all logical operators across codebases.
We already have logic to check logicals - so it seems like a logical extension to add these checks to remove more unnecessary conditionals.

We also already check `||=` and `&&=` - and these checks could also handle the assignment operator variants (eg `num ||= 0`, `bool &&= true`, `maybeNum ??= undefined` are all unnecessary).

I could see the `+` case split out into a new rule like `no-unnecessary-addition-subtraction` or something (because it could also cover `a - 0` and `a -= 0`).

**Comment by Zamiell** [comment]
> It's just about the general pattern of a RHS value that equals the negative case (for || and ??) or the positive case (for &&)...We also already check ||= and &&=

I guess you could make the case that `||=` and `&&=` are comparisons, but my new rule rule also covers the `=` operator, which is clearly not a comparison operator. 

For example, my new rule throws an error on this code:

```ts
declare let foo: true;
foo = true; // Error
```

It also works with enums:

```ts
enum Foo {
  Value1 = "Value1",
  Value2 = "Value2",
}
declare let foo: Foo.Value1;
foo = Foo.Value1; // Error
```

**Comment by Zamiell** [comment]
> (because it could also cover a - 0 and a -= 0).

Right. Currently my rule throws an error with the following operators:

```ts
const USELESS_OPERATORS_WITH_ZERO: ReadonlySet<string> = new Set([
  "+",
  "-",
  "|",
  // "<<" is technically a useless operator, but we whitelist it so that bit flag enums will not
  // cause false positives.
  ">>",
  ">>>",
]);
```

**Comment by bradzacher** [comment]
Yeah I'd separate the two concerns - put the logical stuff into `no-unnecessary-condition`, and put the assignment stuff into a new rule.

A lot of people like having big rules that do a lot as it means there's fewer things to config. I think people care less about unnecessary (re)assignments compared to unnecessary conditions - so people are less likely to turn on a rule just about unnecessary assignments.

Some people also might not care about doing unnecessary assignments - eg ensuring a variable is initialised. TS also can be wrong about types of variables sometimes and definite assignments can complicate things.
